### ISSUE TITLE
Embark keymap prompter needs a few improvements

### ISSUE THREAD

**Issue by minad** [oantolin/embark#186]
Since I am trying to use your Embark prompter now as general prefix command help command, there are a few issues. This is not unexpected I guess, since the Embark keymaps are somehow well-formed and now I am throwing more weird stuff at it.

* Simple menu items are not supported (desc . cmd). Throws an error as of now. This would be nice for Consult Narrowing.
* Complex menu items are probably also not supported (I haven't tested this)
* Consult narrowing keys use these which-key pseudo keys, the Embark prompter thows an error. These bindings should be ignored somehow.

EDIT: I can also provide patches to fix this. I just want to document the things somewhere which do not yet work for me.

**Comment by oantolin** [comment]
Yeah, let's make this robust and general purpose. As you said, needing to work only for Embark keymaps means lots of edge cases had never been tested before.

**Comment by oantolin** [comment]
Ah, I figured out how you are testing this. You have a consult narrow key set! (I do too, it just took a while to click.)

**Comment by minad** [comment]
Using this patch works for me:

~~~~ diff
diff --git a/embark.el b/embark.el
index 25fc7be..db7e557 100644
--- a/embark.el
+++ b/embark.el
@@ -634,7 +634,8 @@ first line of the documentation string; otherwise use the word
-   (or
-    (when (symbolp cmd) (symbol-name cmd))
-    (when-let ((doc (documentation cmd)))
-      (save-match-data
-        (when (string-match "^\\(.*\\)$" doc)
-          (match-string 1 doc))))
-    "<unnamed>")))
+   (cond
+    ((stringp (car-safe cmd)) (car cmd))
+    ((symbolp cmd) (symbol-name cmd))
+    ((when-let (doc (and (functionp cmd) (documentation cmd)))
+       (save-match-data
+         (when (string-match "^\\(.*\\)$" doc)
+           (match-string 1 doc)))))
+    (t "<unnamed>"))))
@@ -653,0 +655,3 @@ first line of the documentation string; otherwise use the word
+                   ;; Filter which-key pseudo keys
+                   ;; TODO more general filter?
+                   unless (eq (car-safe cmd) 'which-key)
~~~

**Comment by oantolin** [comment]
Excellent, thanks a lot! I was on my way to basically the same patch so I'll just use yours.

**Comment by oantolin** [comment]
Weird. No errors anymore but when I use this with the consult narrow key in `consult-buffer`, it only shows the binding `SPC consult-narrow`. Shouldn't more keys be bound to `consult-narrow`.

**Comment by minad** [comment]
Weird. When I am using the Embark patch https://github.com/oantolin/embark/commit/9e2c04f9bc21b7ab0f0dee19ccfa1c56c8d3147b it works for me and shows a list of `consult-narrow` commands with the keys `f, b, ...`. If I also add the Consult patch, https://github.com/minad/consult/commit/70b795e948a52499722c244708db84d1d22f09cf, I also get nice descriptions. I don't know what you are observing.

**Comment by oantolin** [comment]
Oh, it works with Selectrum! I was display the completion with Embark. I must be deduping candidates somewhere. 😮

**Comment by oantolin** [comment]
I haven't update Consult yet, to see the nice descriptions, let me try that too.

**Comment by minad** [comment]
> I must be deduping candidates somewhere.

Uh, probably your prompter needs some tofus :laughing:

**Comment by oantolin** [comment]
Huh, not just Embark, but the `*Completions*` buffer and icomplete too, only show `SPC consult-narrow`. This is weird, I'd say it's compeltion-all-sorted-completions that's deduping but I think the `*Completions*` buffer doesn't go through that function (may it does, I'm not really sure).

**Comment by minad** [comment]
It is justified to bind the same command to multiple keys, but maybe it is less justified that the same command behaves differently then. Maybe I should rework the `consult-narrow` design, hmm.

**Comment by oantolin** [comment]
I'm not a huge fan of commands that use the key they are bound too, but it is not that uncommon in Emacs, so I think you don't you need to change anything.

**Comment by minad** [comment]
As far as I know, default completion/Icomplete performs deduping. Selectrum does not do this, since it considers the deduping unnecessary. Selectrum basically considers it a bug if a completion table has duplicates. You need tofus :worried:

**Comment by oantolin** [comment]
For example `self-insert-command` and `digit-argument` do that.

**Comment by minad** [comment]
Btw, my patch is incomplete. I didn't test the prompter sufficiently. When selecting a candidate I also get an error now for consult narrowing. How does your prompter work, such that it handles different keys? Does it push unread-command-events?

**Comment by oantolin** [comment]
Well, I currently display the key bindings in text properties, so that I can just reuse the `command` completion category, but I could include the key bindings in the candidates (which would also allow matching against them, which sometimes people want), and make up a new completion category.

**Comment by oantolin** [comment]
> Btw, my patch is incomplete. I didn't test the prompter sufficiently. When selecting a candidate I also get an error now for consult narrowing. How does your prompter work, such that it handles different keys? Does it push unread-command-events?

Oh, it doesn't handle keys at all! 😆 It just returns the command you select. If it is bound to several keys, that distinction is lost in the return value.

**Comment by oantolin** [comment]
It's basically completely unprepared for the same command being bound to multiple keys and doing something that depends on the key it is bound to.

**Comment by minad** [comment]
> Well, I currently display the key bindings in text properties, so that I can just reuse the command completion category, but I could include the key bindings in the candidates (which would also allow matching against them, which sometimes people want), and make up a new completion category.

That's probably a better option. I saw there was some longer discussion before regarding this, but I didn't follow it, since at that time I just used the which-key keymap prompter as you know.

> It's basically completely unprepared for the same command being bound to multiple keys and doing something that depends on the key it is bound to.

Okay, I suggest we reopen the issue :)

**Comment by oantolin** [comment]
> Keymap prompter does not seem to handle the nil completion correctly, wrong type argument commandp nil...

Could you explain how you triggered this one? I thought I was filtering out keys bound to nil in `embark--all-bindings`. (Or am I misunderstanding the issue? Is it not about keys bound to nil?)

**Comment by minad** [comment]
> Could you explain how you triggered this one? I thought I was filtering out keys bound to nil in embark--all-bindings. (Or am I misunderstanding the issue? Is it not about keys bound to nil?)

This one is weird. `completing-read` allows to return the empty string in any case even if you require a match. I don't understand why it allows that, but the `completing-read` API is ~crazy~ baroque, so why not? You can trigger it in Selectrum by moving the point to the prompt with no input and press RET. It also works with default completion if you directly press RET without entering any input.

~~~ elisp
(completing-read "Test: " '(a b c) nil 'require-match)
~~~

**Comment by oantolin** [comment]
Gotcha! Thanks.

**Comment by oantolin** [comment]
I fixed a couple more of the bugs listed here:

- Compute maximum width of key bindings, for alignment.
- Handle simple menu items (I'm not actually sure those are menu items, but things of the form (description . command).

Finally, I didn't quite ensure that command are invoked from their key sequence, since that would require inversion of control. (Basically that would involve setting `unread-command-events` and letting the command loop call the command; which would be practical for the new `embark-prefix-help-command`, but not so much for `embark-act`; if I do switch to using `set-transient-map`, then I'd also make this change at the same time.)

Instead I just set `last-command-event` to the last character in the key sequence. This seems to be enough to get `consult-narrow`, `digit-argument` and `self-insert-command` to work, which seems good enough for now.

In particular consult-narrow does work now!

**Comment by minad** [comment]
Thanks for looking more into this. I am not sure if this is due to my setup/misconfigurtion, but the completing read prompter does not work anymore. It is nicely aligned now, sweet! But the commands are not invoked. I tested `C-x C-h ...`, Consult narrowing and Embark actions with `C-h ...`.

**Comment by minad** [comment]
> Handle simple menu items ---I'm not actually sure those are menu items, but things of the form (description . command).

I know, you said that before. This is Emacs terminology. See https://www.gnu.org/software/emacs/manual/html_node/elisp/Simple-Menu-Items.html.

> The simpler (and original) way to define a menu item is to bind some event type (it doesn’t matter what event type) to a binding like this: (item-string . real-binding)

**Comment by oantolin** [comment]
> Thanks for looking more into this. I am not sure if this is due to my setup/misconfigurtion, but the completing read prompter does not work anymore.

It doesn't work for me anymore either! 😆 I believe did test all the changes yesterday, but I didn't try restarting Emacs until now. I'll fix this.

> Do you plan to go back to inversion of control again btw? It seems to me that there is no other way. If we really want to fully support keymaps, we have to give them full control, this means we have to do that.

I do, it's just that I've been putting it off until I have more time, since there are tons of fiddly details. But I think it would fix various different issues that have accumulated, so it's probably a good idea to do it.

**Comment by oantolin** [comment]
Wow, that was a silly mistake! 357a54d11b7d5348fe65ce4e31e56f33a7e34e12

It should work now, including consult-narrow.

**Comment by minad** [comment]
Okay, it works (mostly). There is one thing which does not work. `consult-narrow` does not only use `last-command-event`, but also `(this-single-command-keys)`. This does not work since we don't have the inversion of control.

1. Press `< f`, narrows to file (works)
2. Press `< C-h`, select something (works)
3. Press `< C-h`, select `SPC All` (does not work, narrows to hidden buffers)

This is due to having SPC overloaded for both hidden buffers and to reset the narrowing. It is a bit of a mess since one cold also expect `< SPC` to narrow to hidden buffers. Narrowing to hidden buffers works only if you press SPC as first input, which is okay since I don't see the need to have hidden things to be more accessible.

**Comment by oantolin** [comment]
Is SPC really a good choice for widening? I mean, I understand what you say about not minding that it overrides narrowing to hidden buffers, but why have the widening overlap with any of the default narrowing categories? How about repeating the narrow key instead?

**Comment by minad** [comment]
Repeating the key is also a good idea, maybe I could change the default to that. It is configurable via `consult-widen-key`. Right now the default setting is `consult-narrow-key SPC`.

**Comment by oantolin** [comment]
I do think repeating the key is a nicer default. And it helps to cover up over this deficiency in Embark, until I invert control (back to the way it used to be...).

**Comment by minad** [comment]
Yes, it is a nicer default. Now it is also possible to press `< SPC` and get narrowing to the hidden buffers this way.

**Comment by oantolin** [comment]
> this allows the command to be used in the style of helm/counsel-descbinds

How do you call those commands? From top level and they include *every* keybinding? Or do they prompt for a keymap name?

**Comment by minad** [comment]
I think it just includes every global command. This is nice since one can search through everything which has been bound. It is like describe-bindings.

**Comment by oantolin** [comment]
Ah, OK.

**Comment by oantolin** [comment]
I checked off the two items that mention @ and #192.

**Comment by karthink** [comment]
Is it possible to change the prompt string in the completing-read-prompter from "Command: " to "Command on %s: ", with %s replaced by `(embark--truncate-target ...)`?

I suppose similar adjustments are needed when using the prompter to become a new command.

**Comment by minad** [comment]
@karthink Instead of changing the prompt string we have the minimal indicator. I guess that's good enough?
